src/patterns.md

```diff
@@ -0,0 +1,646 @@
+#     ChangeColor(u8, u8, u8),
+# }
+# let message = Message::Quit;
+match message {
+    Message::Quit => println!("Quit"),
+    Message::WriteString(write) => println!("{}", &write),
+    Message::Move{ x, y: 0 } => println!("move {} horizontally", x),
+    Message::Move{ .. } => println!("other move"),
+    Message::ChangeColor { 0: red, 1: green, 2: _ } => {
+        println!("color change, red: {}, green: {}", red, green);
+    }
+};
+```
+
+## Refutability
+
+A pattern is said to be *Refutable* when it **has the possibility of not being matched**
+by the value it is being matched against. *Irrefutable* patterns, on the other hand,
+always match the value they are being matched against. Examples:
+
+```rust
+let (x, y) = (1, 2);               // "(x, y)" is an irrefutable pattern
+
+if let (a, 3) = (1, 2) {           // "(a, 3)" is refutable, and will not match
+    panic!("Shouldn't reach here");
+} else if let (a, 4) = (3, 4) {    // "(a, 4)" is refutable, and will match
+    println!("Matched ({}, 4)", a);
+}
+```
+
+## Literal patterns
+
+> **<sup>Syntax</sup>**\
+> _LiteralPattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [BOOLEAN_LITERAL]\
+> &nbsp;&nbsp; | [CHAR_LITERAL]\
+> &nbsp;&nbsp; | [BYTE_LITERAL]\
+> &nbsp;&nbsp; | [STRING_LITERAL]\
+> &nbsp;&nbsp; | [RAW_STRING_LITERAL]\
+> &nbsp;&nbsp; | [BYTE_STRING_LITERAL]\
+> &nbsp;&nbsp; | [RAW_BYTE_STRING_LITERAL]\
+> &nbsp;&nbsp; | `-`<sup>?</sup> [INTEGER_LITERAL]\
+> &nbsp;&nbsp; | `-`<sup>?</sup> [FLOAT_LITERAL]
+
+[BOOLEAN_LITERAL]: tokens.html#boolean-literals
+[CHAR_LITERAL]: tokens.html#character-literals
+[BYTE_LITERAL]: tokens.html#byte-literals
+[STRING_LITERAL]: tokens.html#string-literals
+[RAW_STRING_LITERAL]: tokens.html#raw-string-literals
+[BYTE_STRING_LITERAL]: tokens.html#byte-string-literals
+[RAW_BYTE_STRING_LITERAL]: tokens.html#raw-byte-string-literals
+[INTEGER_LITERAL]: tokens.html#integer-literals
+[FLOAT_LITERAL]: tokens.html#floating-point-literals
+
+_Literal patterns_ match exactly the value they represent. Since negative numbers are
+not literals in Rust, literal patterns also accept an optional minus sign before the
+literal.
+
+Floating-point literals are currently accepted, but due to the complexity of comparing
+them, they are going to be forbidden on literal patterns in a future version of Rust (see
+[issue #41620](https://github.com/rust-lang/rust/issues/41620)).
+
+Literal patterns are always refutable.
+
+Examples:
+
+```rust
+for i in -2..5 {
+    match i {
+        -1 => println!("It's minus one"),
+        1 => println!("It's a one"),
+        2|4 => println!("It's either a two or a four"),
+        _ => println!("Matched none of the arms"),
+    }
+}
+```
+
+## Identifier patterns
+
+> **<sup>Syntax</sup>**\
+> _IdentifierPattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `ref`<sup>?</sup> `mut`<sup>?</sup> [IDENTIFIER] (`@` [_Pattern_] ) <sup>?</sup>
+
+_Identifier patterns_ bind the value they match to a **previously undeclared** variable.
```

Comment:
The phrase "previously undeclared" here is confusing to me. I'm not certain if the original intent is to emphasize that the identifier must be unique within a pattern (i.e. `(x, x)` is an invalid pattern), or if it is trying to express something else. The identifier can shadow bindings of the same name from outside the pattern, so that's why I'm confused by the "previously undeclared" part. What do you think of changing it to something like this?

>    _Identifier patterns_ bind the value they match to a variable. The identifier must be unique within the pattern. The variable will shadow any variables of the same name in scope.

Reply:
SGTM. I would also add that the scope of the variable is dependent upon the context of where the pattern is used.